### ISSUE TITLE
detrending should happen on each season rather than the whole timeseries

### ISSUE THREAD

**Issue by apendergrass** [apendergrass/cmip6hack-precip#28]
sd_hist = (xr.apply_ufunc(signal.detrend, data_tmp['hist'].fillna(0),
                                    kwargs={'axis': 0}).where(~data_tmp['hist'].isnull())
               .sel(time=data_tmp['hist'].time.dt.season.isin(seas))
               .std("time"))
    sd_futr = (xr.apply_ufunc(signal.detrend, data_tmp['futr'].fillna(0),
                                    kwargs={'axis': 0}).where(~data_tmp['futr'].isnull())
               .sel(time=data_tmp['futr'].time.dt.season.isin(seas))
               .std("time"))

i think detrending is applied to the whole timeseries for each month, but the seasonal mean should be calculated first, then the timeseries of seasonsal means should be detrended, then the std should come last

**Comment by apendergrass** [comment]
started fixing this on branch fixstdcalc, have something in a new branch, also not sure the master is working right now...

**Comment by apendergrass** [comment]
- this is mostly ok, just the years need to be right now

**Comment by chiaral** [comment]
> ```
> sd_hist = (xr.apply_ufunc(signal.detrend, data_tmp['hist'].fillna(0),
>                                 kwargs={'axis': 0}).where(~data_tmp['hist'].isnull())
>            .sel(time=data_tmp['hist'].time.dt.season.isin(seas))
>            .std("time"))
> sd_futr = (xr.apply_ufunc(signal.detrend, data_tmp['futr'].fillna(0),
>                                 kwargs={'axis': 0}).where(~data_tmp['futr'].isnull())
>            .sel(time=data_tmp['futr'].time.dt.season.isin(seas))
>            .std("time"))
> ```
> 
> i think detrending is applied to the whole timeseries for each month, but the seasonal mean should be calculated first, then the timeseries of seasonsal means should be detrended, then the std should come last

This is a bit of a long string of commands, so hard to follow, but yes i think you are right. I guess you could simply move the 
```.where(~data_tmp['hist'].isnull())
           .sel(time=data_tmp['hist'].time.dt.season.isin(seas)```
inside. But there are better ways to do it.

**Comment by chiaral** [comment]
this
```
    if seas=="DJF":
        dec = data_tmp['hist']['pr'].sel(time=data_tmp['hist'].time.dt.month.isin(12))[0:-1]
        jan = data_tmp['hist']['pr'].sel(time=data_tmp['hist'].time.dt.month.isin(1))[1:]
        feb = data_tmp['hist']['pr'].sel(time=data_tmp['hist'].time.dt.month.isin(2))[1:]
        timeseries_of_seasonalmeans = (np.array(dec)+np.array(jan)+np.array(feb))/3
    elif seas=="JJA":
        jun = data_tmp['hist']['pr'].sel(time=data_tmp['hist'].time.dt.month.isin(6))[1:]
        jul = data_tmp['hist']['pr'].sel(time=data_tmp['hist'].time.dt.month.isin(7))[1:]
        aug = data_tmp['hist']['pr'].sel(time=data_tmp['hist'].time.dt.month.isin(8))[1:]
        timeseries_of_seasonalmeans = (np.array(jun)+np.array(jul)+np.array(aug))/3

```

can be swapped with `.time.dt.season.isin(seas)`
xarray has the season time attribute. I could work on it next week if you want.

**Comment by chiaral** [comment]
(sorry for the multiple messages..)
The way you do it now it's problematic for time series that could have not homogeneous length. 
if you do xarray.time.dt.season.isin('DJF') you will be assured that you are not mixing DJF among years, or having to deal with truncated time series (which I guess it is very unlikely in model, more likely in observations). 
My understanding is that you calculate a time series of DJF and then dretrend that. so for 10 years of data you detrend 10 values of DJF (correct?) so something like:

       ` DJF_timeseries = data_tmp['hist']['pr'].sel(time=data_tmp['hist'].time.dt.season.isin('DJF').mean('time')`

should give you that.

**Comment by apendergrass** [comment]
Chiara, thanks for having a look at this and helping me learn xarray. I got confused and couldn't figure out how to do the thing that needs to be done with xarray commands, which is why I switched to brute-force explicit numpy. 

Tinkering around locally, I figured out that .time.dt.season.isin(seas) just makes a timeseries of monthly values, instead of first taking the seasonal means. Then it was detrending the monthly values, instead of first taking the seasonal mean, so it was not consistent with the original figure. 

I figured out something using xarray stuff. it's more complicated and still needs an if statement. I guess I also like for it to be really clear what's happening in code when possible, so I know things aren't wrong, and can double-check long after I wrote it without breaking it into pieces and rerunning. 

Here is the new one: 

if seas == "DJF":
    middle_month = 1
elif seas == "JJA":
    middle_month = 7
    
rollingmean = data_tmp['hist']['pr'].sel(time=data_tmp['hist'].time.dt.season.isin(seas)).rolling(time=3,min_periods=3,center=True).mean()
timeseries_of_seasonalmeans = rollingmean.sel(time=rollingmean.time.dt.month.isin(middle_month))

sd_hist = xr.apply_ufunc(signal.detrend, timeseries_of_seasonalmeans.dropna('time'),kwargs={'axis': timeseries_of_seasonalmeans.dims.index('time')}).std("time")

I didn't add yet because when I tested it, the answers were weird... the first period should be close to zero change (bc it has a lot of overlap with the base period - 1976-2005 versus 1991-2020), but they're pretty different. 

Still need to go back and get the December before the first year for the DJFs, to be consistent with previous code (but that shouldn't have a big effect)

**Comment by apendergrass** [comment]
Ok I realized what was going on in the middle of the night (classic) - the future only included data from SSPs, so 2006 onward, so the variability was super low for only 15 instead of 30 years. 

Now there's a step where hist and ssps are concatenated to fix this too. 

I feel ok about this now (except the reconciling DJF stuff, but that's in a separate issue), and you can close this issue if you're ok with it too.

**Comment by chiaral** [comment]
Hi. My bad. Indeed there is another way to do it. Group by and then mean. I can refactor it next week if you are interested. But everything works really. 

Best. Sure close it! 
C

**Comment by apendergrass** [comment]
Nbd! This way works and there are other things that don’t yet, so let’s do those - I’ll close it
